Move specialist document pages

leenagupte opened the pull request:
, [Jira issue PNP-6373](https://gov-uk.atlassian.net/browse/PNP-6373)⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Copy the rendering code for specialist document pages from government-frontend.
The rendering code has been modified to fit the best practices for frontend as outlined in the Contributing guide.

## Why

[Trello card](https://trello.com/c/RrUeLTKc)

## Examples

aaib-report:
* https://www.gov.uk/aaib-reports/aaib-investigation-to-rotorsport-uk-calidus-g-pcpc
* https://govuk-frontend-app-pr-4600.herokuapp.com/aaib-reports/aaib-investigation-to-rotorsport-uk-calidus-g-pcpc

business-finance-support:
* https://www.gov.uk/business-finance-support/rural-business-growth-fund
* https://govuk-frontend-app-pr-4600.herokuapp.com/business-finance-support/rural-business-growth-fund

cma-case:
* https://www.gov.uk/cma-cases/topps-tiles-slash-ctd-tiles-certain-assets-merger-inquiry
* https://govuk-frontend-app-pr-4600.herokuapp.com/cma-cases/topps-tiles-slash-ctd-tiles-certain-assets-merger-inquiry

maib-report:
* https://www.gov.uk/maib-reports/capsize-of-recreational-craft-wheelyboat-123-with-the-loss-of-2-lives
* https://govuk-frontend-app-pr-4600.herokuapp.com/maib-reports/capsize-of-recreational-craft-wheelyboat-123-with-the-loss-of-2-lives

eu-withdrawal-act-2018-statutory-instrument:
* https://www.gov.uk/eu-withdrawal-act-2018-statutory-instruments/the-drivers-hours-and-tachographs-amendment-regulations-2021
* https://govuk-frontend-app-pr-4600.herokuapp.com/the-drivers-hours-and-tachographs-amendment-regulations-2021

N.B. These have a link to the finder at the bottom of the page.

## Screenshots

|government-frontend (before) | frontend (after)|
|-------------------------------|----------------|
|![screencapture-gov-uk-business-finance-support-access-to-finance-advice-north-west-england-2025-02-17-13_55_13](https://github.com/user-attachments/assets/2086dcee-b016-46fc-9989-33499adf6c09)|![screencapture-frontend-dev-gov-uk-business-finance-support-access-to-finance-advice-north-west-england-2025-02-17-14_20_09](https://github.com/user-attachments/assets/6ac01e99-7b76-45bf-8fef-0a5b4e6083eb)|
|![Screenshot 2025-02-17 at 13 56 58](https://github.com/user-attachments/assets/a84b3246-5bc9-4344-a141-04c079e66396)|![Screenshot 2025-02-17 at 13 57 35](https://github.com/user-attachments/assets/b7032fa3-c0fc-446a-9a51-b25f84fd0bf1)|

## Testing in integration

|Rendered view|Content item|
|---------------|-------------|
|<img width="960" alt="Screenshot 2025-02-24 at 17 38 32" src="https://github.com/user-attachments/assets/8b763245-2020-4a68-bfb0-bc17300e48e9" />|<img width="820" alt="Screenshot 2025-02-24 at 17 37 58" src="https://github.com/user-attachments/assets/a36761c9-49de-4185-b5ab-25de1bc33572" />|

leenagupte committed:
Add specialist document examples to dev controller and README

leenagupte committed:
Add placeholder routes and methods for specialist document

Specialist documents need to use the "Full path routing constraint" as
their routes are comprised of `/<document_type>/slug` rather than just
`/slug`

leenagupte committed:
Copy specialist document view from government-frontend

This copies the view exactly as it is in government-frontend.
Subsequent commits will update the view to use values from the
ContentItem model and presenter.

Audit trail:
specialist_document.html.erb: https://github.com/alphagov/government-frontend/blob/a7e99cf556dbc5d8b8fab5a317157136cadeb8a0/app/views/content_items/specialist_document.html.erb

leenagupte committed:
Fix lint errors in specialist document show template

Runs `bundle exec erb_lint --lint-all --autocorrect` to fix the errors

leenagupte committed:
Add SpecialistDocument model with continuation link details

Audit trail:

specialist_document_presenter.rb: https://github.com/alphagov/government-frontend/blob/a7e99cf556dbc5d8b8fab5a317157136cadeb8a0/app/presenters/specialist_document_presenter.rb#L35-L47

leenagupte committed:
Update headers in specialist document template

leenagupte committed:
Update continuation link details in specialist document view

leenagupte committed:
Add headers to SpecialistDocument model

The list of headers is used to generate the contents list if there are
more than 3 headers in the list.

Confusingly this list of headers is generated in multiple places.

[specialist-publisher] uses the govspeak gem to parse the body content
and look for `h2` and `h3` tags and generate the `headers` list in
`details` from them. If there are no `h2` or `h3` tags the entire
`headers` element is removed from `details.

government-frontend deals with `headers` in two different ways. It uses
the `headers` element in details to [create the contents list], but then
somewhat replicates the code from the govspeak gem by [parsing the `h2`
tags] in the body content to determine whether or not so show the
contents list.

Those it's not being used to render the contents list in
government-frontend the `headers` element does have a `level` attribute
that indicates whether the header is a `h2` or `h3`. Therefore it should
not be possible to greatly simplify the code from its current form in
government-frontend.

The first step is to just extract the headers list in the
SpecialistPublisher model. The code to determine whether or not to
display the contents list will be added to a presenter in a later
commit.

An example from content schemas is being used to initialise the
SpecialistDocument model object, however  the ["details"]["headers"]
have been copied from the example into the test as unlike other model
tests the values in the array are being modified and the test is easier
to follow if you can see exactly what's changed.

Audit trail:
specialist_document_presenter: https://github.com/alphagov/government-frontend/blob/c6eead39b29192e692c802c9b5e8bc234e16692c/app/presenters/specialist_document_presenter.rb#L86-L100

[specialist-publisher](https://github.com/alphagov/specialist-publisher/blob/efadf3c4df33d045c0067e095f958d4b12638f29/app/presenters/document_presenter.rb#L50)
[create the contents list]: https://github.com/alphagov/government-frontend/blob/c6eead39b29192e692c802c9b5e8bc234e16692c/app/presenters/specialist_document_presenter.rb#L86-L100
[parsing the `h2` tags]: https://github.com/alphagov/government-frontend/blob/main/app/presenters/content_item/contents_list.rb#L43

leenagupte committed:
Add show_finder_link? method to SpecialistDocumentPresenter

Audit trail:
specialist_document_presenter.rb: https://github.com/alphagov/government-frontend/blob/c6eead39b29192e692c802c9b5e8bc234e16692c/app/presenters/specialist_document_presenter.rb#L49-L53

leenagupte committed:
Add specialist document css

Audit trail:
_specialist_document.scss: https://github.com/alphagov/government-frontend/blob/89f73ddbf4d5596e920c5f21f8b18e3db3b04899/app/assets/stylesheets/views/_specialist-document.scss
dartsass.rb: https://github.com/alphagov/government-frontend/blob/89f73ddbf4d5596e920c5f21f8b18e3db3b04899/config/initializers/dartsass.rb#L13

leenagupte committed:
Add protection type to specialist document model

leenagupte committed:
Add protection type image details to specialist document model

This is to get the images when the document type is
"protected_food_drink_name".

It might be worth separating this into a separate model and presenter.

leenagupte committed:
Add show_protection_type_image? to specialist document presenter

leenagupte committed:
Add protection image files

leenagupte committed:
Add protection_image_path method

leenagupte committed:
Add protection_image_alt_text method

leenagupte committed:
Add translations for protection image alt text

leenagupte committed:
Update how govspeak is called in specialist document view

leenagupte committed:
Update how description is called from specialist document view

leenagupte committed:
Use initial publication date in specialist documents view

Audit trail:

specialist_document.html.erb: https://github.com/alphagov/government-frontend/blob/868603abc5c2e3a50cd78dbb505e56ed3102f5a1/app/views/content_items/specialist_document.html.erb#L61
updatable.rb: https://github.com/alphagov/government-frontend/blob/main/app/presenters/content_item/updatable.rb#L3-L5

leenagupte committed:
Use updated and history in specialist document view

This uses the Updatable concern and published_dates component to display
the updated date and change history in the view.

Audit trail:

specialist_document.html.erb: https://github.com/alphagov/government-frontend/blob/868603abc5c2e3a50cd78dbb505e56ed3102f5a1/app/views/content_items/specialist_document.html.erb#L62-L63
updatable.rb: https://github.com/alphagov/government-frontend/blob/868603abc5c2e3a50cd78dbb505e56ed3102f5a1/app/presenters/content_item/updatable.rb#L7-L15

leenagupte committed:
Display publisher metadata specialist documents view

This partial was added an renamed in a previous PR.

Audit trail:
specialist_document.html.erb: https://github.com/alphagov/government-frontend/blob/a718ddf50ab81320b3586bfb31acf0a9ea7ef82e/app/views/content_items/specialist_document.html.erb#L20

leenagupte committed:
Add specialist document request tests

leenagupte committed:
Add specialist document system tests

Audit trail:
specialist_document_test.rb: https://github.com/alphagov/government-frontend/blob/main/test/integration/specialist_document_test.rb
test_helper.rb: https://github.com/alphagov/government-frontend/blob/main/test/test_helper.rb#L78-L146

leenagupte committed:
Limit updates shown for specialist documents

This replicates the logic from government-frontend so that the
conditions under which the change history is shown to the user remain
unchanged.

Audit trail:

specialist_document_presenter.rb: https://github.com/alphagov/government-frontend/blob/8799336fd3c96dac7c7e53cce3cd8a522beb17da/app/presenters/specialist_document_presenter.rb#L219-L227

KludgeKML commented on app/views/specialist_document/show.html.erb:
```diff
@@ -10,7 +10,10 @@
 <div class="govuk-grid-row">
   <div class="govuk-grid-column-two-thirds responsive-top-margin">
-    <%= render "govuk_publishing_components/components/title", @content_item.title_and_context %>
+    <%= render "govuk_publishing_components/components/title", {
```
I don't know if this is covered in a later commit, but we don't want to be using the title component - we want to be using the [heading component[(https://components.publishing.service.gov.uk/component-guide/heading) instead. (title is being removed very soon).

KludgeKML commented on app/views/specialist_document/show.html.erb:
```diff
@@ -10,10 +10,13 @@
-        title: content_item.title,
-        average_title_length: "long",
-      } %>
+    <%= render "govuk_publishing_components/components/heading", {
```
Ah, there we are. I think we could just do this in one commit.

KludgeKML commented on app/presenters/specialist_document_presenter.rb:
```diff
@@ -0,0 +1,19 @@
+  end
+
+  def level_two_headings?
+    content_item.headers.any? { |header| header[:level] == 2 }
```
Can we make any assumptions about the header levels? If the roots are always level 2, then we could just replace this with `.any?`, couldn't we? (since we're nesting the lower level ones). Or if the roots are sometimes level 1, this isn't going to find the level 2 ones if they're nested.

leenagupte replied:
The `headers` are created by parsing markdown added by editors. So I don't think we can make any assumptions about the header levels. 

The code that this logic is based on checks for the existence of ["h2" elements in the body text](https://github.com/alphagov/government-frontend/blob/main/app/presenters/content_item/contents_list.rb#L43). I've tried to do the same thing, but in a nicer way and use the `headers` object we access to.

leenagupte replied:
> roots are sometimes level 1, this isn't going to find the level 2 ones if they're nested

If there are multiple H1s there are bigger problems with the docs. But it does mean that a contents list won't be created if there are just H3s.

KludgeKML commented on app/models/specialist_document.rb:
```diff
@@ -19,6 +19,7 @@ def headers_list(headers)
       h = {
         href: "##{header['id']}",
         text: header["text"].gsub(/:$/, ""),
+        level: header["level"],
```
Maybe put this line in the previous commit?

KludgeKML reviewed (commented):
Looks pretty good, I have a few inline notes.

leenagupte committed:
Add SpecialistDocumentPresenter

This adds a `contents` method. The method needs to know if
there are any level two headers. Rather than parsing the html in the
body attribute, the `level` is just being acquired from the `headers`
list in the model.

There is a lot of code in the original government-frontend code relating
to images, and the character count and the number of table rows. This
seems to be more related to document collections, rather specialist
documents, so for now it has been omitted and rather than creating a
shared presenter, a specialist document specific one has been created
to minimise the code.

Audit trail:

contents_list_presenter.rb: https://github.com/alphagov/government-frontend/blob/main/app/presenters/content_item/contents_list.rb#L21

leenagupte commented:
@KludgeKML Thanks for your review. I've amended the commits as suggested, and responded to your question about the headers.

deborahchua commented on app/views/specialist_document/show.html.erb:
```diff
@@ -38,7 +38,7 @@
     <% end %>
 
-    <%= render "components/contents_list_with_body", contents: @content_item.contents do %>
+      <%= render "components/contents_list_with_body", contents: @specialist_document_presenter.contents do %>
```
Minor: an additional space here

leenagupte replied:
Good spot! I've updated and fixed up.

deborahchua reviewed (approved):
This looks good! I appreciate the small commits, it's made reviewing it really easy 🎉

leenagupte committed:
Model a finder content item

All specialist documents are linked to a finder. To display the
facet information for a specialist document we need to cross-reference
the facets selected for the document with the full facet details from
the finder.

Rather than model the data from the finder in the specialist documents
model, a separate model has been created. Finders are a type of content
item, so by doing this we get the `base_path` attribute for free. Also
the model will be reused when finder pages are finally consolidated into
frontend.

leenagupte committed:
Add metadata to specialist document model

Audit trail:
specialist_document_presenter.rb: https://github.com/alphagov/government-frontend/blob/1503b2907f39e2f11881d5a6c66644792022d549/app/presenters/specialist_document_presenter.rb#L110-L119

leenagupte commented:
@KludgeKML @deborahchua Thank you both for your reviews. After our conversations about modelling organisations I have changed this PR slightly to create a Finder content item model, and use that to model the `finder` from `links`. It doesn't change the code very much as I refrained from going one step further and left `facets` (from `details`) as an array of hashes rather than an array of open structs.

I think this approach is better, but I'd appreciate a re-review.

leenagupte committed:
Add facet_values method to specialist document model

This method takes the facets the content item is mapped to from the
metadata, and the full details of all of the facets from
["links"]["finders"] and returns one list that only contains the full
details of the facets the document is mapped to.

The specialist document presenter and view will need to know certain
details to determine how to render the facet details in the metadata
section. For example, if the facet is filterable a link to the finder
will be rendered. Rather than just passing "filterable" to the
presenter to sort out, the filterable is being used to determine
whether a facet is just text or is just a list of links and instead
adjusts the value for type. This allows us to limit the amount of
information that needs to be processed in the presenter.

KludgeKML commented on app/views/specialist_document/show.html.erb:
```diff
@@ -10,7 +10,13 @@
-    <%= render "govuk_publishing_components/components/title", @content_item.title_and_context %>
+    <%= render "govuk_publishing_components/components/heading", {
+      text: content_item.title,
+      average_title_length: "long",
```
I think you still want to remove this? (this is valid for the title component, not for the heading one)

leenagupte replied:
Good spot! Thanks. I copy and pasted from the case study view 🙈 , which will also need to be updated (in a separate change)

leenagupte replied:
I've made the change and fixed up.

KludgeKML commented on app/controllers/specialist_document_controller.rb:
```diff
@@ -1,5 +1,7 @@
 
-  def show; end
+  def show
+    @specialist_document_presenter = SpecialistDocumentPresenter.new(content_item)
```
:+1:

(a bunch of existing examples of this code have content_item.to_h, we should probably clean them up)

KludgeKML commented on app/models/specialist_document.rb:
```diff
@@ -1,10 +1,12 @@
   def initialize(content_store_response)
     super(content_store_response)
 
     @continuation_link = content_store_hash.dig("details", "metadata", "continuation_link")
```
Note - content_store_hash may or may not exist depending on if this is merged before or after the PfG scaffolding PR (Nothing necessarily to do here, just need to note that these two PRs will clash one way or the other, and we have to bear in mind that some rebasing will be needed)

KludgeKML commented on app/presenters/specialist_document_presenter.rb:
```diff
@@ -40,4 +61,23 @@ def protected_food_drink_name?
     content_item.document_type == "statutory_instrument"
   end
+
+  def format_facet_value(type, value, key, filterable)
+    if type == "date"
+      view_context.display_date(value)
+    elsif type == "text" && value.is_a?(Array) && filterable == true
+      value.map { |v| format_filterable_links(key, v[:value], v[:label]) }
+    else
+      value
+    end
+  end
```
I wonder if we could do this mapping in the view? (Just because this is the first time in the new consolidated routes that we've passed a view_context to a presenter rather than it just being available in views and helpers, so want to double-check whether this is a legit precedent to set).

Current precedent for it being in view: https://github.com/alphagov/frontend/pull/4225/files#diff-0b7fa2198812d14b54b691c4fde2f55cab0ba0109aaf95efaffba7e58e587d39R4-R10

leenagupte replied:
I went back and forth about this. I'm not happy about passing in a `view_context` but then I didn't want to duplicate the helper methods either. 

The reason I didn't put it directly in the view is because of the amount of logic in it, especially in the `elsif`. That `format_facet_value` method is called inside a loop and it would get messy and hard to debug.

I also felt it would be easier to test in a unit test rather than a system test.

leenagupte replied:
I'd be happy to pair on a better way to do this, as I'm sure there must be one.

leenagupte commented on app/presenters/specialist_document_presenter.rb:
```diff
@@ -40,4 +61,23 @@ def protected_food_drink_name?
+    end
+  end
+
+  def format_filterable_links(key, value, label)
```
Do you mean `_link` singular? i.e. `format_filterable_link`?

leenagupte committed:
Use heading component in specialist document view

Changes the view to call the heading component rather than the title
component, and passes values to the component directly, rather than via a
presenter method.

Audit trail:

specialist_document_presenter.rb: https://github.com/alphagov/government-frontend/blob/96d0c2416e31816e84a4f49a15904bc7098b7960/app/presenters/specialist_document_presenter.rb#L12
specialist_document.html.erb: https://github.com/alphagov/government-frontend/blob/a718ddf50ab81320b3586bfb31acf0a9ea7ef82e/app/views/content_items/specialist_document.html.erb#L12

leenagupte committed:
Add important_metadata presenter method

The view_context is being passed to the presenter from controller.
An array of selected facet values is returned from the model. Rather
than looping through the model in the view to apply the helper methods
it was was felt better to move the complexity to a presenter which is
easier to test than a view. The compromise is that rather than
recreating the date and link helper methods, they are instead being
accessed in presenter via the view_context. There's no need to define
a new view_context as the controller already has access to it.

The choice to pass in the view_context is a compromise. A helper method
could have been created instead. However, helpers are global, and
therefore should be generic so that they can be used by multiple views.
But the facet logic is specialist document specific, so really should
be in a presenter. Helpers and Presenters are both [presentational], so
this feels like an acceptable compromise to avoid duplicating the
helper methods and the reduce the amount of complexity.

If a helper was created, then the logic gets more complex as the facet
still needs to have the filtered finder link applied in the presenter,
but the "type" would  also need to be passed to the helper so it knows
whether to style the facet as a link, a date, or not style it at all.
This requires the presenter to return a nested array of hashes. What
this means is that both the presenter and the helper would need to loop
through the facets for little gain.

Audit trail:

specialist_document_presenter.rb: https://github.com/alphagov/government-frontend/blob/main/app/presenters/specialist_document_presenter.rb#L27
specialist_document_presenter_test.rb: https://github.com/alphagov/government-frontend/blob/868603abc5c2e3a50cd78dbb505e56ed3102f5a1/test/presenters/specialist_document_presenter_test.rb

[presentational]: https://blog.railsforgedev.com/rails-view-patterns-helpers-vs-partials-vs-presenters-vs-decorators#heading-when-to-use-helpers

KludgeKML commented on app/helpers/link_helper.rb:
```diff
@@ -16,7 +16,7 @@ def govuk_styled_link(text, path: nil, inverse: false)
     "<a href='#{path}' class='#{classes}'>".html_safe + text + "</a>".html_safe
   end
```
Isn't this last commit now covered in https://github.com/alphagov/frontend/pull/4660/files ?

leenagupte replied:
Yes, I added it to get the tests to pass, but will remove it and rebase.

leenagupte committed:
Show contents list in specialist document view

This updates the view to use the new `contents` method from
SpecialistDocumentPresenter.

leenagupte committed:
Show finder link in specialist document view

This updates the view to use the new `show_finder_link?` method from
SpecialistDocumentPresenter, and base_path and title from the modelled
finder content item.

leenagupte committed:
Use protection image methods in view

leenagupte committed:
Use important_metdata method in view

KludgeKML reviewed (approved):
LGTM!

AshGDS reviewed (commented):
Thanks for providing page examples across live/the PR 👍 

I've done a HTML code compare rather than a visual check as a first pass, and noticed some stuff:

1. It seems the description meta tag has gone missing between the live site and the preview app? e.g. `<meta name="description" content="Location: Roadford Lake, Devon, England.">`

2. There seems to be an empty div just below the h1 on the pages - `<div class="govuk-grid-column-one-third"></div>` - is this needed?

3. There's a `direction-ltr` class on the `cma-case` page which looks like it doesn't do anything (though I'm not sure you've added this in the code, I can't actually find where it's coming from)

4. Having the language set on the `<main>` element is used for GA4 tracking purposes. Is it possible for the language to be included on the main tag? Our tracking does fall back to the `<html>` language value if it can't find it on the `<main>` element, but that one isn't always reflective of the content language.

JamesCGDS reviewed (commented):
I’ve checked the following pages:

- https://www.gov.uk/business-finance-support/access-to-finance-advice-north-west-england
- https://www.gov.uk/cma-cases/review-of-banking-for-small-and-medium-sized-businesses-smes-in-the-uk
- https://www.gov.uk/international-development-funding/uk-aid-direct
- https://www.gov.uk/european-structural-investment-funds/business-development
- https://www.gov.uk/service-standard-reports/apply-for-a-budgeting-loan

I only noticed one visual difference on < tablet screen sizes where additional spacing has been introduced above the “Related Content” section (screenshot below). This is because in this app, the [shared sidebar navigation component](https://github.com/alphagov/frontend/blob/main/app/views/shared/_sidebar_navigation.html.erb) (which contains “Related Content”) is wrapped in a `div` with a `sidebar-navigation` class which adds `margin-top: govuk-spacing(7);` on smaller screens.

Personally I don’t I think this is a problem as I feel the extra spacing gives things a bit more room to breathe but happy to discuss if others feel differently.

| ##Before | ## After |
|-----------|---------|
| ![image](https://github.com/user-attachments/assets/06177da3-d45b-4eb6-b61e-f5ba080f9c28) | ![image](https://github.com/user-attachments/assets/c4dd7134-fe88-4672-878f-73606d118785) |

leenagupte committed:
Add meta description to specialist document pages

The machine readable metadata component only adds the `og` tag
i.e. `<meta name="og:description" content="description">`

It doesn't add `<meta name="description" content="description"> to the
Head.

It seems that all of content item pages are adding this tag in their
templates rather having it appear once in the layout so specialist
documents are following the same pattern for now.

leenagupte commented:
@AshGDS Thanks for being so thorough in your review. ⭐ 

I'll address each of the points you raised:

1. Good spot on the meta description. I've added it in this [commit](https://github.com/alphagov/frontend/pull/4600/commits/890f28786fa3c0a43e1c01441b8adbd7938c857c).
2. I've added a [PR](https://github.com/alphagov/frontend/pull/4661) to remove the empty `<div>` tag if there aren't any available translations.
3. The `direction-ltr` is on all pages, even the homepage 😱 so I shall leave it as it is. I suspect it's being pulled in from the layout somewhere.
4. I thought the language was being added to the `<main>` tag? `<main>` is being set in the [application layout](https://github.com/alphagov/frontend/blob/main/app/views/layouts/application.html.erb#L16) so fixing it probably falls outside of the scope of this change.

leenagupte commented:
@JamesCGDS Thanks for your review 🪇 

That's a good spot about the `side-navigation`. As it's something that's applied globally, it's probably something that should be discussed, and then decide whether it should be changed.

AshGDS reviewed (approved):
LGTM 👍 

I guess with the language thing, since it's populated by the content item I assume it will work OK once deployed then?

And James is now on leave, so with the spacing thing he noticed it may be worth running it by Nikin for approval.

leenagupte commented:
> LGTM 👍
> 
> I guess with the language thing, since it's populated by the content item I assume it will work OK once deployed then?
> 
> And James is now on leave, so with the spacing thing he noticed it may be worth running it by Nikin for approval.

Thanks for re-reviewing!

It looks like the language thing isn't working in frontend at all. As it's supposed to be added in the layout, I think it should be fixed in a separate PR. The same goes for the spacing issue James found. That's also something that's being set in the layout, so should also have a separate discussion and fix. I can raise them as issues.
